### PR TITLE
When updating member states after a profile change, skip custom profiles in rooms.

### PR DESCRIPTION
**Pull request by Half-Shot**
This change is so that member events with custom profile info will stick between updates of a users profile. Currently this checks the old member profile against the old profile and uses that to determine whether it should be updated. This might turn out to be too flimsy so any hole poking is appreciated.

Signed-off-by: Will Hunt <half-shot@molrams.com>

### CODE REVIEW

**Comment by erikjohnston** [comment]
@matrixbot half-shot is a nice guy so its ok to test

**Comment by Half-Shot** [comment]
uuuuuhuh. Not quite sure how https://github.com/matrix-org/sytest/blob/develop/tests/30rooms/30history-visibility.pl#L375 is now failing :/

**Comment by erikjohnston** [comment]
That test has started become flakey, its on my todo list to sort out

**Comment by Half-Shot** [comment]
Alrighty, then I'll just ignore that one for now. When you've got a second, could you look through this and see if there is anything wrong with the PR itself?

File: synapse/handlers/profile.py [hunk 1]
```diff
@@ -162,18 +166,32 @@ def _update_join_states(self, requester):
 
         for room_id in room_ids:
             handler = self.hs.get_handlers().room_member_handler
+            member_event = yield handler.get_member_event(user, room_id)
```
Comment (richvdh) [inline synapse/handlers/profile.py]: I believe you want `StateHandler.get_current_state` for this. See https://github.com/matrix-org/synapse/blob/develop/synapse/api/auth.py#L123 for example.

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -177,6 +177,18 @@ def update_membership(
         defer.returnValue(result)
 
     @defer.inlineCallbacks
+    def get_member_event(self, target, room_id):
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: as above, probably don't need this

**Comment by Half-Shot** [comment]
Changes made. 
Now looking at @erikjohnston like https://cloud.githubusercontent.com/assets/2072976/24215816/e11c02ea-0f31-11e7-9dda-48448d03b6d3.jpg

**Comment by erikjohnston** [comment]
The tests are faaiiiiiiling.

I'm a bit worried that if the display names get out of sync we won't have any way of just resetting *all* the display names. This is particularly annoying because the updating of the memberships events happen one at a time, so confusion reigns if someone sets their display name twice in rapid succession.

**Comment by Half-Shot** [comment]
>I'm a bit worried that if the display names get out of sync we won't have any way of just resetting all the display names. This is particularly annoying because the updating of the memberships events happen one at a time, so confusion reigns if someone sets their display name twice in rapid succession.

Yeah this worries me too. I have a few options:

1. Custom members should have an additonal key that says "leave me alone"
2. Standard member events should do the inverse of the above, which might be more sensible.
3. Synapse keeps a table of what local member events it is allowed to modify, which is extra ick.
4. Do nothing, and sad things happen when I try to update my profile.

**Comment by erikjohnston** [comment]
> 3. Synapse keeps a table of what local member events it is allowed to modify, which is extra ick.

This may be best, in terms of having a separate API to set profiles per room.

All of this sounds like it needs to be thought about and/or go through the spec :s

**Comment by erikjohnston** [comment]
Closing for now pending discussion on exactly how we want to handle the previous cases

**Pull request closed without merging**